### PR TITLE
Only wrap descendants with render method, closes #84

### LINKED ISSUE #84: Appears to wrap other functions than createClass()
As reported in https://github.com/gaearon/react-transform-catch-errors/issues/25, it seems that the plugin regressed in 2.x and thinks `React.shape` is something like `React.createClass` just because it receives an object as the argument.

### PR DESCRIPTION
**Pull request by nfcampos**
closes #84 (see the issue for a discussion of the changes)

## todo
- [x] fix current tests
- [x] make change to index.js
- [x] write additional tests

### CODE REVIEW

**Comment by nfcampos** [comment]
@gaeron should i write a new test for each of the cases the current tests used to cover?

**Comment by gaearon** [comment]
Please do.

**Comment by nfcampos** [comment]
something like this is not supposed to be wrapped, is it?
```js
import React, { Component } from 'react'
class Foo extends Component {
  render() {}
}
```

**Comment by gaearon** [comment]
This looks like a React component to me, why not?

**Comment by nfcampos** [comment]
because it requires changing this line https://github.com/gaearon/babel-plugin-react-transform/blob/master/src/index.js#L174
to `superClasses: options.superClasses || ['React.Component', 'Component'],
which mean that any class that extends a class called `Component` and has a render method is wrapped.
is that ok? or is there some other way of doing this?

**Comment by gaearon** [comment]
Good catch. I would say let’s add `Component` to defaults. It’s a too common pattern.

**Comment by nfcampos** [comment]
yeah i agree. does that go into this PR or another one (it's not really related to this)?

**Comment by gaearon** [comment]
It’s fine to do this here.

**Comment by nfcampos** [comment]
Alright, it's all done i think.
changes:
- only wrap classes that extend the correct superClass *and* have a render method
- changed tests to reflect change above
- added `Component` to default superClasses
- added test for the above

**Comment by gaearon** [comment]
Seems like technically this wouldn’t even be a breaking change because we used to allow any class with `render()` anyway.

**Comment by nfcampos** [comment]
well it breaks some stuff that shouldn't work anyways, so that part is fine

to me it looks like the only breaking change would be stuff like this:
```js
import React from 'react'
const Comp = React.Component
class Foo extends Comp {
  render() {}
}
```
this would have worked previously because it had a render method and now won't, but would anyone actually do something like this?

**Comment by gaearon** [comment]
Even if they used this pattern, we’re not technically “breaking” it because the code would still work. Transforms are meant for development only anyway.

**Comment by nfcampos** [comment]
yeah that's true

File: src/index.js [hunk 1]
```diff
@@ -119,7 +119,7 @@ export default function({ types: t, template }) {
     CallExpression(path) {
       if (
         path.node[VISITED_KEY] ||
-        !matchesPatterns(path.get('callee'), this.factoryMethods) &&
+        !matchesPatterns(path.get('callee'), this.factoryMethods) ||
```
Comment (jwbay) [inline src/index.js]: if both `matchesPatterns` and `isReactLikeComponentObject` are independent checks for "is this node a react component", then we should only bail and return early if *both* the checks fail, right? if a component would pass the `isReactLikeComponentObject` check but fails the `factoryMethods` check, it never gets a chance to do so:
`if (false || true || false)` -> bail

**Comment by gaearon** [comment]
Out in 2.0.1, thank you a lot!

**Comment by nfcampos** [comment]
Happy to help!

**Comment by gaearon** [comment]
If you have time, I’d appreciate you looking into why https://github.com/gaearon/babel-plugin-react-transform/pull/87 fails, merging/updating it if it’s correct, and checking whether it fixes another correctness issue: https://github.com/gaearon/babel-plugin-react-transform/issues/88.

**Comment by nfcampos** [comment]
I'll look into it

**Pull request merged**